### ISSUE TITLE
how to use lstm for classification with keras

### ISSUE THREAD

**Issue by trangtv57** [fchollet/keras#6084]
I have X have dimension: (30,3) x: [[1.2, 0.2, 0.6],[0.6, 0, 0.9],....[0.6, 0, 0.8]] and y just [0,1,0, .1], i read about sequence to sequence with keras in this post :https://blog.keras.io/building-autoencoders-in-keras.html, but i can't understand ( in Sequence-to-sequence autoencoder). I think i need use TimeDistributed layer in keras for X then use this to input for LSTM ( many to many), sry about question, but i try read to much, but can't clear to understand how resolve my problem.

**Comment by aryopg** [comment]
LSTM accept 3D input, and (i assume that X is the input) the dimension of the X is only 2D. `expand_dim` would solve that problem. And i don't really understand why do you need to use `seq2seq` or `TimeDistributed` cause if your question is to do a classification, you can just add a simple `Dense` layer at the end of the model

**Comment by trangtv57** [comment]
sry, i ask this question in 1 forum and they suggest me find out about sequence to sequence, so if i want to set time_step by hand, how can i use expand_dim with x. (i think TimeDistributed to gen dataset each timestep ). And the last, after first hidden layer is LSTM, can i use or should i use another LSTM layer, or Dense layer. thanks you.

**Comment by aryopg** [comment]
I don't really understand your question/problem, so correct me if i interpret your question wrongly. 

Sequence-to-sequence model is a model that take sequence as an input and sequence as an output. As you might has seen in some blog post or forum, the applications of sequence-to-sequence architecture are related to generating another sequence (machine translation, chatbot). I think the reason why some people want you to take a look at a sequence-to-sequence model and TimeDistributed layer is (probably) because they want you to take a look at the **attention mechanism** . Have you heard about it before? If you haven't heard about it, i suggest you to take a look at [this paper](https://arxiv.org/pdf/1409.0473.pdf) first. And then probably you can take a look at [this issue](https://github.com/fchollet/keras/issues/4962).

Regarding about you wanting to add another layer of LSTM/Dense on top of your architecture, it is all up to you, if you want to add LSTM again, make sure that the LSTM layer(before the one that you recently add) return sequence (just add `return_sequence=True`). And then, if you want to add a Dense layer, which i think necessary in order to do a classification task, the last LSTM layer should not return_sequence.

I hope this will help you :)

**Comment by trangtv57** [comment]
thanks you very much, i have someday to read more about time step in lstm, with LSTM layer in keras it has 3 dimension (batchsize, timestep, n_Feature) so if my input have just 2 dimension like in my question, so how can i expand_dim or something to my feature have 3 dimesion to it can contain timestep. Because i just use expand_dim or reshape function in numpy, but it just can convert to np array have 1 timestep. :(

**Comment by aryopg** [comment]
base on what i saw on your first question, i think it can only have 1 timestep. May i know the real form of the data so that i can help to figure out how to generate timesteps?

**Comment by trangtv57** [comment]
my data is in https://github.com/defcom17/NSL_KDD, i use it for trainning myself, with 41 feature, and almost 100000 sample ( 100000 vector each vector has 41 feature) and label is [0,1] 
[[0 0 17 9 491 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 2 0.0 0.0 0.0 0.0 1.0  0.0 0.0 150 25 0.17 0.03 0.17 0.0 0.0 0.0 0.05 0.0]
 [0 1 42 9 146 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 13 1 0.0 0.0 0.0 0.0 0.08 0.15 0.0 255 1 0.0 0.6 0.88 0.0 0.0 0.0 0.0 0.0]
 [0 0 47 5 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 123 6 1.0 1.0 0.0 0.0 0.05  0.07 0.0 255 26 0.1 0.05 0.0 0.0 1.0 1.0 0.0 0.0]
 [0 0 21 9 232 8153 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 5 5 0.2 0.2 0.0 0.0 1.0  0.0 0.0 30 255 1.0 0.0 0.03 0.04 0.03 0.01 0.0 0.01]
 [0 0 21 9 199 420 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 30 32 0.0 0.0 0.0 0.0  1.0 0.0 0.09 255 255 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0]
 [0 0 47 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 121 19 0.0 0.0 1.0 1.0 0.16  0.06 0.0 255 19 0.07 0.07 0.0 0.0 0.0 0.0 1.0 1.0]
 [0 0 47 5 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 166 9 1.0 1.0 0.0 0.0 0.05  0.06 0.0 255 9 0.04 0.05 0.0 0.0 1.0 1.0 0.0 0.0]
 [0 0 47 5 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 117 16 1.0 1.0 0.0 0.0 0.14  0.06 0.0 255 15 0.06 0.07 0.0 0.0 1.0 1.0 0.0 0.0]
 [0 0 49 5 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 270 23 1.0 1.0 0.0 0.0 0.09  0.05 0.0 255 23 0.09 0.05 0.0 0.0 1.0 1.0 0.0 0.0]
 [0 0 47 5 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 133 8 1.0 1.0 0.0 0.0 0.06  0.06 0.0 255 13 0.05 0.06 0.0 0.0 1.0 1.0 0.0 0.0]]
this is X[0:10]. thanks

**Comment by trangtv57** [comment]
I just try to reshape my train data like : x(125920,41) -> x(1574,80,41) so time step now is 80. and with y_train i use reshape too y(12590) to y(15784, 80)
this topo run:
  model = Sequential()
    model.add(LSTM(120, input_shape=(80, 41), return_sequences=True))
    # model.add(Dropout(0.5))
    model.add(LSTM(80))
    # model.add(Dropout(0.5))
    model.add(Dense(1))
but result not like i expected.
I change y from 1 dimension just 1/0 to 5 dimension ( with 1 hot encoding sklearn : 0 [1, 0, 0, 0 ,0] 1: [0, 1, 0, 0, 0] so y_train  has shape: (1574,80,5) so in final Dense layer i change to 
 model.add(Dense((80,.5)))
but Its error, can you tell me why and how to fix it guys :(

**Comment by potholiday** [comment]
@aryopg  I think @trangtv57  ment to say that his **X** a sequence is in the shape of (30,3) and **y** is in the shape of (1,4) ie for four classes then how should we pass **X** and **y** to lstm. See shape of X and y dosent match **X** has 30 rows  any **y** has one

**Comment by trangtv57** [comment]
thanks for reply @potholiday :)
the problem you talk to represent my problem is not my purpose i want to ask, can skip number of row X, y. I want focus on reshape my X from [..,3] mean 3 columns and n row( n record) to shape can train in LSTM (with time series not equal 1)

**Comment by potholiday** [comment]
@trangtv57  Sorry but still I am not getting your question? Do you want to reshape your data to pass it through lstm? That is from 2d to 3d, say you have (100,2) shaped data to (100,1,2) data?

**Comment by trangtv57** [comment]
yes @potholiday  I want to know what is time step mean, and if I want to train with my data: (100,2), with time step is 7, how can i do it?, and does more time step is good than 1 ?

**Comment by potholiday** [comment]
you have to reshape your input data.. This is a code i find in the internet https://stackoverflow.com/a/38084402/996366 . Please go through the code

**Comment by trangtv57** [comment]
thanks you @potholiday. I will try your suggest.

**Comment by potholiday** [comment]
@trangtv57  I was trying out that example in the link. some how it dosent work when time step is greater than 1. I have found this example https://stackoverflow.com/a/43274707/996366 I dont know how accurate is this?
